src/mame/brother/lw30.cpp

```diff
@@ -0,0 +1,1105 @@
+	int incr = m_incr;
+
+	if(m_frequency > 0)
+		clock = m_frequency;
+
+	/* if we're not enabled, just fill with 0 */
+	if(m_state == 0xff || clock == 0)
+	{
+		buffer.fill(0);
+		return;
+	}
+
+	/* fill in the sample */
+	for(int sampindex = 0; sampindex < buffer.samples(); sampindex++)
+	{
+		if(BIT(m_state, 7) != 0)
+			buffer.put(sampindex, signal > 0 ? 1.f : -1.f);
+		else
+			buffer.put(sampindex, 0.f);
+
+		incr -= clock;
+		while(incr < 0)
+		{
+			incr += rate;
+			signal = -signal;
+		}
+	}
+
+	/* store progress through wave */
+	m_incr = incr;
+	m_signal = signal;
+}
+
+WRITE_LINE_MEMBER(brother_beep_device::set_state)
+{
+	// only update if new state is not the same as old state
+	if(m_state == state)
+		return;
+
+	m_stream->update();
+
+	if(m_state == 0) {
+		// restart wave from beginning
+		m_incr = 0;
+		m_signal = 0x01;
+	}
+	m_state = state;
+}
+
+void brother_beep_device::set_clock(uint32_t frequency)
+{
+	if(m_frequency == frequency)
+		return;
+
+	m_stream->update();
+	m_frequency = frequency;
+	m_signal = 0x01;
+	m_incr = 0;
+}
+
+class lw30_state : public driver_device
+{
+public:
+	lw30_state(const machine_config &mconfig, device_type type, const char *tag)
+		: driver_device(mconfig, type, tag),
+		maincpu(*this, "maincpu"),
+		screen(*this, "screen"),
+		floppy(*this, "floppy"),
+		beeper(*this, "beeper"),
+		m_io_kbrow(*this, "kbrow.%u", 0),
+		rom(*this, "maincpu"),
+		font_normal(*this, "font_normal"),
+		font_bold(*this, "font_bold")
+	{ 
+		//video_control = 0b00000010; // TEST LW-10 screen height
+	}
+
+	void lw30(machine_config& config);
+
+private:
+	// devices
+	required_device<hd64180rp_device> maincpu;
+	required_device<screen_device> screen;
+
+	required_device<floppy_connector> floppy;
+	required_device<brother_beep_device> beeper;
+	optional_ioport_array<9> m_io_kbrow;
+	required_region_ptr<uint8_t> rom, font_normal, font_bold;
+
+	// floppy
+	uint8_t floppy_data{};
+	uint8_t io_88{};
+	uint8_t io_98{};
+	uint8_t floppy_control{}; // stepper motor control
+	uint8_t floppy_steps{}; // quarter track
+	uint8_t floppy_shifter{}, floppy_latch{};
+	bool floppy_read_until_zerobit{};
+
+	// video
+	uint8_t videoram[0x2000]; // 80 chars * 14 lines; 2 bytes per char (attribute, char)
+	uint8_t video_cursor_x, video_cursor_y, video_pos_x, video_pos_y, video_control, io_b8;
+	uint8_t cursor_state;
+
+	// screen updates
+	uint32_t screen_update(screen_device& screen, bitmap_rgb32& bitmap, const rectangle& cliprect);
+
+	uint8_t illegal_r(offs_t offset, uint8_t mem_mask = ~0) {
+		LOG("%s: unmapped memory read from %0*X & %0*X\n", machine().describe_context(), 6, offset, 2, mem_mask);
+		return 0;
+	}
+	void illegal_w(offs_t offset, uint8_t data, uint8_t mem_mask = ~0) {
+		LOG("%s: unmapped memory write to %0*X = %0*X & %0*X\n", machine().describe_context(), 6, offset, 2, data, 2, mem_mask);
+	}
+
+	// ROM
+	uint8_t rom42000_r(offs_t offset, uint8_t mem_mask = ~0) {
+		return rom[0x02000 + offset] & mem_mask;
+	}
```

Comment:
You don’t need to use `mem_mask` like this – you only need to check it if lane select or tristate drive is important.  You can omit the parameter altogether if you only use this as a memory access handler.  It’s usually unused by 8-bit CPUs for memory accesses anyway as lane select within a byte is almost unheard-of.